packages/server/src/time-control.ts

```diff
@@ -40,100 +54,326 @@ export function ValidateTimeControlBase(
+  constructor() {
+    this._timeoutService = getTimeoutService();
+  }
+
+  initialState(
+    variant: string,
+    config: IConfigWithTimeControl,
+  ): TimeControlParallel {
+    const numPlayers = makeGameObject(variant, config).numPlayers();
+
+    const timeControl: TimeControlParallel = {
+      moveTimestamps: [],
+      forPlayer: {},
+    };
+
+    for (let i = 0; i < numPlayers; i++) {
+      timeControl.forPlayer[i] = {
+        remainingTimeMS: config.time_control.mainTimeMS,
+        onThePlaySince: null,
+        stagedMoveAt: null,
+      };
+    }
+
+    switch (config.time_control.type) {
+      case TimeControlType.Absolute: {
+        // nothing to do
+        break;
+      }
+
+      case TimeControlType.Fischer: {
+        // nothing to do
+        break;
+      }
+
+      default: {
+        console.error("received config with invalid time control type");
+      }
+    }
+
+    return timeControl;
+  }
+
   handleMove(
-    _game: GameResponse,
-    _game_obj: AbstractGame<unknown, unknown>,
-    _playerNr: number,
-    _move: string,
-  ): Promise<void> {
-    console.log(
-      "time control handler for parallel moves is not implemented yet",
-    );
-    return;
+    game: GameResponse,
+    game_obj: AbstractGame<unknown, unknown>,
+    playerNr: number,
+  ): ITimeControlBase {
+    const config = game.config as IConfigWithTimeControl;
+
+    switch (config.time_control.type) {
+      case TimeControlType.Absolute: {
+        let timeControl = game.time_control as TimeControlParallel;
+        // could happen for old games
+        if (timeControl === undefined) {
+          timeControl = this.initialState(game.variant, config);
+        }
+
+        const timestamp = new Date();
+
+        timeControl.moveTimestamps.push(timestamp);
+        timeControl.forPlayer[playerNr].stagedMoveAt = timestamp;
+        this._timeoutService.clearPlayerTimeout(game.id, playerNr);
+
+        // check if the round finishes with this move
+        //
+        // I'm not sure if this will always work as intended
+        // because this is the game object after the move has been played
+        // whereas we actually need the players that are on the play in the round that the move belongs to
+        if (
+          game_obj
+            .nextToPlay()
+            .every(
+              (player_nr) =>
+                timeControl.forPlayer[player_nr].stagedMoveAt !== null,
+            )
+        ) {
+          for (const player_nr of game_obj.nextToPlay()) {
+            const playerData = timeControl.forPlayer[player_nr];
+
+            // time control starts only after first move of player
+            if (
+              game.moves.filter((move) => player_nr in move).length >
+              (player_nr === playerNr ? 0 : 1)
+            ) {
+              playerData.remainingTimeMS -=
+                playerData.stagedMoveAt.getTime() -
+                playerData.onThePlaySince.getTime();
```

Comment:
I think there's a problem here. A player can change their staged move, right? So here's a scenario:
1. Player A stages a move and has 10 seconds left. Player B didn't stage a move yet and has 30 seconds left.
2. 20 seconds pass.
3. Player A changes their mind and stages a new move. They now have a remaining time of -10 seconds.

So player A should either be prevented from changing their move, if they don't have enough time, or the remaining time should be calculated base on the first time a player stages a move in a new round.

Reply:
> So player A should either be prevented from changing their move

FWIW I'm okay with preventing players from changing their move altogether if it helps.  That was one of the things I wasn't sure about when implementing Parallel Go.

Reply:
right, thank you for finding that. I'm unsure how to handle that situation right now, but I'll think about how to fix it.

Reply:
Changed such that an error is thrown if changing a staged move would cause the players time to go to zero or below.